docs/developing-fuzzbench/adding_a_new_benchmark.md

```diff
@@ -161,18 +212,80 @@ export FUZZER_NAME=afl
+
+Building benchmarks and fuzzers entails the following process:
+
+1. The benchmark image is built. This image is defined by
+   `benchmarks/$BENCHMARK/Dockerfile`. It inherits from
+   `gcr.io/oss-fuzz-base/base-builder` which provides clang and other things
+   needed by benchmarks (particular OSS-Fuzz benchmarks to build). Standard
+   benchmarks (usually) inherit from the latest version of `base-builder`
+   while OSS-Fuzz benchmarks (usually) inherit from the specific version of
+   `base-builder` that was used to build the version of the project's source
+   (commit) that the benchmark uses. This is to ensure that builds of these
+   benchmarks just work and don't break when `base-builder` is updated to use a
+   new version of clang. Note that pinning some benchmarks to specific versions
+   of clang is a bit ugly and this behavior may change in the future.
+
+1. The fuzzer builder image is built. This image is defined by
+   `fuzzers/$FUZZER/builder.Dockerfile`. This dockerfile will inherit from
+   a parent image that is provided to it at buildtime (using the docker
+   variable: `parent_image`). The parent image provided is the benchmark docker
+   image from the previous step. The fuzzer builder image builds the fuzzer sets
+   up anything the fuzzer needs to build the benchmark (such as `FUZZER_LIB`).
+
+1. The benchmark builder image is built. This image is defined by
+   `docker/benchmark-builder/Dockerfile`. This inherits from the fuzzer builder
+   image. This is the first image in this build process that is defined by the
+   main FuzzBench code (e.g. not fuzzers, benchmarks, or OSS-Fuzz). Its first
+   function is to copy the FuzzBench code and install packages needed to run
+   FuzzBench like Python3.7 For benchmarks that define a `commit` in their
+   `benchmark.yaml` (i.e. OSS-Fuzz benchmarks) the build process for this image
+   checks out the source code of that project at the specified commit. Then the
+   process defines the environment variables `CC`, `CXX`, `CXXFLAGS`, `CFLAGS`
+   and `OUT`. It then calls the `build` function from the fuzzer's `fuzzer.py`
+   file. `build` can change these environment variables as needed and then calls
+   `build_benchmark` from `fuzzers/utils.py`. `build_benchmark` invokes the
+   `build.sh` file of the benchmark building it with environment variables
+   provided by `fuzzer.py` or the benchmark builder image. `build` then copies
+   build of the `fuzzer` (e.g. `afl-fuzz`) to `$OUT` (one reason why we do this
+   here instead of when building the fuzzer is because the `build.sh` can
+   overwrite it, in the future we will probably ensure that the build processes
+   for the benchmark and the fuzzer don't interfere). In some cases, such as
+   QSYM (which is no longer a supported fuzzer for other reasons), `build` can
+   reset `OUT` so that it can build the benchmark twice since some fuzzers may
+   need two different builds of the same benchmark.
```

Comment:
Should we mention the intermediate-runners (fuzzers/$FUZZER/runner.Dockerfile)?

Reply:
I was planning on putting that in the "Running section" but I think that would be confusing. I'm leaving this as a TODO. Frankly it's somwhat awkward to put in this file because runners is pretty nonrelevant for benchmark integrators. I think in a future version I will move this section somewhere else and link to it from here.

Reply:
SG